File: assignment2/Oyas-Rijo-Nishan-LakshmyKanth/AVLTree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        self.height = 1 

class AVLTree:
    def get_height(self, node):
        if node is None:
            return 0
        return node.height

    def get_balance_factor(self, node):
        if node is None:
            return 0
        return self.get_height(node.left) - self.get_height(node.right)

    def right_rotate(self, y):
        x = y.left
        temp = x.right

        x.right = y
        y.left = temp

        y.height = 1 + max(self.get_height(y.left), self.get_height(y.right))
        x.height = 1 + max(self.get_height(x.left), self.get_height(x.right))

        return x 

    def left_rotate(self, x):
        y = x.right
        temp = y.left

        y.left = x
        x.right = temp

        x.height = 1 + max(self.get_height(x.left), self.get_height(x.right))
        y.height = 1 + max(self.get_height(y.left), self.get_height(y.right))

        return y

    def insert(self, current_node, new_key):
        if current_node is None:
            return Node(new_key)
        
        if new_key < current_node.data:
            current_node.left = self.insert(current_node.left, new_key)
        elif new_key > current_node.data:
            current_node.right = self.insert(current_node.right, new_key)
        else:
            return current_node 

        current_node.height = 1 + max(self.get_height(current_node.left), 
                                      self.get_height(current_node.right))

        balance = self.get_balance_factor(current_node)

        if balance > 1 and new_key < current_node.left.data:
            return self.right_rotate(current_node)

        if balance < -1 and new_key > current_node.right.data:
            return self.left_rotate(current_node)

        if balance > 1 and new_key > current_node.left.data:
            current_node.left = self.left_rotate(current_node.left)
            return self.right_rotate(current_node)

        if balance < -1 and new_key < current_node.right.data:
            current_node.right = self.right_rotate(current_node.right)
            return self.left_rotate(current_node)

        return current_node
    def show_preorder(self, node):
        if node:
            print(node.data, end=" ")
            self.show_preorder(node.left)
            self.show_preorder(node.right)

my_tree = AVLTree()
root_node = None
numbers = [10, 20, 30, 40, 50, 25]

for n in numbers:
    root_node = my_tree.insert(root_node, n)

print("\nFinal Root:", root_node.data)
print("Tree in Pre-Order:")
my_tree.show_preorder(root_node)
print()

File: Rashmi/avlTree.c
#include <stdio.h>
#include <stdlib.h>
struct node
 {
    int d, h;
    struct node *l, *r;
};
int max(int a, int b) 
{ 
    return a > b ? a : b; 
}
int height(struct node *n) 
{ 
   return n ? n->h : 0; 
}
struct node* newNode(int d) 
{
    struct node* n = malloc(sizeof(struct node));
    n->d = d; n->l = n->r = NULL; n->h = 1;
    return n;
}
struct node* rightRotate(struct node* y) 
{
    struct node* x = y->l;
    y->l = x->r;
    x->r = y;
    y->h = max(height(y->l), height(y->r)) + 1;
    x->h = max(height(x->l), height(x->r)) + 1;
    return x;
}
struct node* leftRotate(struct node* x) 
{
    struct node* y = x->r;
    x->r = y->l;
    y->l = x;
    x->h = max(height(x->l), height(x->r)) + 1;
    y->h = max(height(y->l), height(y->r)) + 1;
    return y;
}
int balance(struct node* n) 
{
    return n ? height(n->l) - height(n->r) : 0;
}
struct node* insert(struct node* r, int d) 
{
    if (!r) return newNode(d);
    if (d < r->d) r->l = insert(r->l, d);
    else if (d > r->d) r->r = insert(r->r, d);
    else return r;
    r->h = 1 + max(height(r->l), height(r->r));
    if (balance(r) > 1 && d < r->l->d) return rightRotate(r);
    if (balance(r) < -1 && d > r->r->d) return leftRotate(r);
    if (balance(r) > 1 && d > r->l->d) {
        r->l = leftRotate(r->l);
        return rightRotate(r);
    }
    if (balance(r) < -1 && d < r->r->d) {
        r->r = rightRotate(r->r);
        return leftRotate(r);
    }
    return r;
}
void inorder(struct node* r) {
    if (r) {
        inorder(r->l);
        printf("%d ", r->d);
        inorder(r->r);
    }
}
int main() 
{
    struct node* root = NULL;
    root = insert(root, 30);
    root = insert(root, 20);
    root = insert(root, 40);
    root = insert(root, 10);
    inorder(root);
}